network/ocsp.go

```diff
@@ -0,0 +1,232 @@
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ocsp"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+const (
+	ocspRequiredYes int = iota
+	ocspRequiredNo
+	ocspRequiredAll
+)
+
+const (
+	ocspFromCertUse int = iota
+	ocspFromCertTrust
+	ocspFromCertPrefer
+	ocspFromCertIgnore
+)
+
+type OCSPConfig struct {
+	url      string
+	required int // ocspRequired*
+	fromCert int // ocspFromCert*
+}
+
+func NewOCSPConfig(uri, required, fromCert string) (*OCSPConfig, error) {
+	if len(uri) > 0 {
+		_, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var requiredVal int
+	switch required {
+	case "yes", "true":
+		requiredVal = ocspRequiredYes
+	case "no", "false":
+		requiredVal = ocspRequiredNo
+	case "all":
+		requiredVal = ocspRequiredAll
+	default:
+		return nil, errors.New("Invalid `ocsp_required` value '" + required + "', should be one of 'yes', 'no', 'all'")
+	}
+
+	var fromCertVal int
+	switch fromCert {
+	case "use":
+		fromCertVal = ocspFromCertUse
+	case "trust":
+		fromCertVal = ocspFromCertTrust
+	case "prefer":
+		fromCertVal = ocspFromCertPrefer
+	case "ignore":
+		fromCertVal = ocspFromCertIgnore
+	default:
+		return nil, errors.New("Invalid `ocsp_from_cert` value '" + fromCert + "', should be one of 'use', 'trust', 'prefer', 'ignore'")
+	}
+
+	if len(uri) > 0 {
+		log.Debugf("OCSP: Using server '%s'", uri)
+	}
+
+	switch required {
+	case "yes", "true":
+		log.Debugf("OCSP: At least one OCSP server should confirm certificate validity")
+	case "no", "false":
+		log.Debugf("OCSP: Allowing certificates not known by OCSP server")
+	case "all":
+		log.Debugf("OCSP: Requiring positive response from all OCSP servers")
+	}
+
+	switch fromCert {
+	case "use":
+		log.Debugf("OCSP: using servers described in certificates if nothing passed via command line")
+	case "trust":
+		log.Debugf("OCSP: trusting responses from OCSP servers listed in certificates")
+	case "prefer":
+		log.Debugf("OCSP: server from certificate will be prioritized over one from command line")
+	case "ignore":
+		log.Debugf("OCSP: ignoring OCSP servers described in certificates")
+	}
+
+	return &OCSPConfig{url: uri, required: requiredVal, fromCert: fromCertVal}, nil
+}
+
+func (c *OCSPConfig) Describe() string {
+	return fmt.Sprintf("url=%s, required=%d, fromCert=%d", c.url, c.required, c.fromCert)
+}
+
+func queryOCSP(commonName string, clientCert, issuerCert *x509.Certificate, ocspServer string) (*ocsp.Response, error) {
```

Comment:
It would be nice to extract this into an interface:
```go
type OcspServer interface {
    Query(commonName string, clientCert, issuerCert *x509.Certificate) *ocsp.Response, error
}
```
This allows separating verification logic from query combination (instead of hardwiring it in `checkOCSP`). More importantly, this enables writing unit tests for OCSP query combianation *and* certificate verification as a whole, because you will be able to write stub OCSP servers that way.